File: maxDepth.cpp

//
// Created by Golfer on 2020/8/7.
//

#include <cstdio>
#include <clocale>
#include<algorithm>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
      int val;
     TreeNode *left;
     TreeNode *right;
      TreeNode(int x) : val(x), left(NULL), right(NULL) {}
  };

// Down-Top
//class Solution {
//public:
//    int maxDepth(TreeNode* root) {
//        if (! root)
//            return 0;
//        int left_max_depth = maxDepth(root->left);
//        int right_max_depth = maxDepth(root->right);
////        return max(left_max_depth, right_max_depth) + 1;
//        return (left_max_depth > right_max_depth) ? left_max_depth + 1 : right_max_depth + 1;
//    }
//};

// Top-Down
class Solution {
public:
    int max_depth = 0;
    int maxDepth(TreeNode* root) {
        get_max(root, max_depth + 1);
        return max_depth;
    }
    void get_max(TreeNode* cur, int depth) {
        if (! cur)
            return;
        if (! cur->left && ! cur->right) {
            max_depth = max(max_depth, depth);
        }
        get_max(cur->left, depth + 1);
        get_max(cur->right, depth + 1);
    }
};

File: maxDepth.py
from typing import List

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# Down-Top
# class Solution:
#     def maxDepth(self, root: TreeNode) -> int:
#         if root is None:
#             return 0
#         # if root.left is None and root.right is None: # No Need in fact
#         #     return 1
#         left_max_depth = self.maxDepth(root.left)
#         right_max_depth = self.maxDepth(root.right)
#         return max(left_max_depth, right_max_depth) + 1

# Top-Down
class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        self.max_depth = 0
        def get_max(node, depth):
            if node is None:
                return
            if node.left is None and node.right is None:
                self.max_depth = max(self.max_depth, depth)
            get_max(node.left, depth+1)
            get_max(node.right, depth+1)
        get_max(root, self.max_depth+1)
        return self.max_depth



